packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +1246,1035 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+/// is called on the associated [controller]. The [OverlayPortal] uses
+/// [overlayChildBuilder] to build its overlay child and renders it on the
+/// closest [Overlay] as if it was inserted using an [OverlayEntry], while it
+/// can depend on the same set of [InheritedWidget]s (such as [Theme]) that this
+/// widget can depend on.
+///
+/// This widget requires an [Overlay] ancestor in the widget tree when its
+/// overlay child is showing. The overlay child will be painted in front of the
+/// [OverlayEntry] closest to this widget in the widget tree (which is usually
+/// the enclosing [Route]). When there are more than one [OverlayPortal] targets
+/// the same [OverlayEntry], the overlay child of the last [OverlayPortal] to
+/// have called `show` appears in the front, regardless of their locations in
+/// the widget tree, or the order in which they were added to the [Overlay].
+///
+/// When [OverlayPortalController.hide] is called, the widget built using
+/// [overlayChildBuilder] will be removed from the widget tree the next time the
+/// widget rebuilds. Stateful descendants in the overlay child subtree may lose
+/// states as a result.
+///
+/// {@tool dartpad}
+/// This example uses an [OverlayPortal] to build a tooltip that becomes visible
+/// when the user taps on the [child] widget. There's a [DefaultTextStyle] above
+/// the [OverlayPortal] controlling the [TextStyle] of both the [child] widget
+/// and the widget [overlayChildBuilder] builds, which isn't otherwise doable if
+/// the tooltip was added as an [OverlayEntry].
+///
+/// ** See code in examples/api/lib/widgets/overlay/overlay_portal.0.dart **
+/// {@end-tool}
+///
+/// ### Differences between [OverlayPortal] and [OverlayEntry]
+///
+/// The main difference between [OverlayEntry] and [OverlayPortal] is that
+/// [OverlayEntry] builds its widget subtree as a child of the target [Overlay],
+/// while [OverlayPortal] uses [overlayChildBuilder] to build a child widget of
+/// itself. This allows [OverlayPortal]'s overlay child to depend on the same
+/// set of [InheritedWidget]s as [OverlayPortal], and it's also guaranteed that
+/// the overlay child will not outlive its [OverlayPortal].
+///
+/// On the other hand, [OverlayPortal]'s implementation is more complex. For
+/// instance, it does a bit more work than a regular widget during global key
+/// reparenting. If the content to be shown on the [Overlay] doesn't benefit
+/// from being a part of [OverlayPortal]'s subtree, consider using an
+/// [OverlayEntry] instead.
+///
+/// See also:
+///
+///  * [OverlayEntry], an alternative API for inserting widgets into an
+///    [Overlay].
+///  * [Positioned], which can be used to size and position the overlay child in
+///    relation to the target [Overlay]'s boundaries.
+///  * [CompositedTransformFollower], which can be used to position the overlay
+///    child in relation to the linked [CompositedTransformTarget] widget.
+class OverlayPortal extends StatefulWidget {
```

Comment:
Is there any value in making this more generic so that it's not tied to Overlay by default?  I can't think of a good example of why you would need a portal to somewhere that's not in an Overlay, but it seems like it wouldn't be too hard to generalize.  React portals are not tied to the idea of an overlay or to the root of the DOM tree or anything, they can be inserted at any DOM element.  Though admittedly modals etc. are always the examples that are used for React portals.

Reply:
This only works on `Overlay` because the properties of its render object allow render children to be added during layout.